canonical/ubuntu-advantage-client
# d/postinst: xenial enable unauthenticated apt source for apps/infra

Allow postinstall to pre-enable visibility to unauthenticate esm-apps and esm-infra repos via adding apt sources  files and Pin-Priority: never in /etc/apt/preferences.d/. 

Also add  python logic to uaclient/entitlements/esm.py for ESMInfra and ESMApps to support re-setting the pin-prirority: never when esm-infra or esm-apps is disabled on xenial.

--- postinst commit
Refactor a couple of functions in ubuntu-advantage-tools.postinst
to a install_esm_apt_key_and_source which will reuse common logic
to emit unauthenticated APT sources.list.d files for both esm-apps
and esm-infra.

Once APT sources are on disk, emit a APT preferences file that pins
the repo origins as Pin-Priority: never which tells apt to list
those packages as upgradeable, but not attempt to install or upgrade
during typical apt-get upgrade|dist-upgrade.

Ran shellcheck on this content and fixed up the suggestions
thoughout the file.

Fixes: #1529

--- uacllient.entitlements. esm.py commit 
    esm: add pin never when disabling esm-infra/apps on xenial via repo_pin_priority and disable_apt_auth_only property overrides


## Proposed Commit Message
Rebase and merge 

## Test Steps
Integration test should cover the artifacts leaving a magically disabled esm-infra/esm-apps APT config on unattached systems.

Otherwise one could;
```bash
sed -i 's/UNRELEASED/xenial/ debian/changelog
git commit -am 'changelog'
build-package
sbuild-it ../out/ubuntu-advantage-tools_27.0.dsc
lxc launch ubuntu-daily:xenial test-x-postinst
lxc file push ubuntu-advantage-tools_27.0_amd64.deb test-x-postinst/

# Should not see esm APT artifacts or policy before install
lxc exec test-x-postinst -- find /etc/apt | grep esm
lxc exec test-x-postinst -- apt policy | grep esm
lxc exec test-x-postinst -- dpkg -i /ubuntu-advantage-tools_27.0_amd64.deb
# Should see esm APT artifacts or policy post install
lxc exec test-x-postinst -- find /etc/apt | grep esm
lxc exec test-x-postinst -- apt policy | grep esm

# check `apt update` output related to ESM Apps when enabled and disabled
apt update..
 82 packages can be upgraded. Run 'apt list --upgradable' to see them.
9 of the updates are from UA Apps: ESM.

```

## Desired commit type::


## Checklist:
